Introduce Management Interface (#30506)
quarkusio/quarkus: docs/src/main/asciidoc/management-interface-reference.adoc

If _https_ is configured, plain HTTP requests will be rejected.

== Configure the root path

By default, management endpoints are exposed under `/q`.
This _root path_ can be configured using the `quarkus.management.root-path` property.
For example, if you want to expose the management endpoints under `/management` use:

[source, properties]
----
quarkus.management.enabled=true
quarkus.management.root-path=/management
----

The mounting rules of the management endpoints slightly differ from the ones used when using the main HTTP server:

* Management endpoints configured using a _relative_ path (not starting with `/`) will be served from the configured root path.
For example, if the endpoint path is `health` and the root path is `management`, the resulting path is `/management/health`
* Management endpoints configured using an _absolute_ path (starting with `/`) will be served from the root.
For example, if the endpoint path is `/health`, the resulting path is `/health`, regardless of the root path
* The management interface does not use the HTTP root path from the main HTTP server.

> == Management endpoints

Reviewer: ```suggestion
== Create a management endpoint
```

> SmallRye Health Checks, SmallRye Metrics and Prometheus are declared as management endpoints.

Reviewer: Can we keep this in terms of extensions (also, combine with / absorb following note). This particular comment should move higher in the doc, as it doesn't just apply to extension writers.

```suggestion
SmallRye Health Checks, SmallRye Metrics, and Micrometer Prometheus will be declared as management endpoints when the management interface is enabled.
```

NOTE: if you do not enable the management interface, these endpoints will be served using the main HTTP server (under `/q` by default).

> To declare a management endpoint, an extension must produce a _non application_ route and call the `management()` method:

Reviewer: ```suggestion
Extensions can create a management endpoint by defining a _non application_ route and calling `management()` method:
```

[source, java]
----
@BuildStep
void createManagementRoute(BuildProducer<RouteBuildItem> routes,
        NonApplicationRootPathBuildItem nonApplicationRootPathBuildItem,
        MyRecorder recorder) {

    routes.produce(nonApplicationRootPathBuildItem.routeBuilder()
        .management() // Must be called BEFORE the routeFunction method
        .routeFunction("my-path", recorder.route())
        .handler(recorder.getHandler())
        .blockingRoute()
        .build());
    //...
}
----

If the management interface is enabled, the endpoint will be exposed on: `http://0.0.0.0:9000/q/my-path`.
Otherwise, it will be exposed on: `http://localhost:8080/q/my-path`.

IMPORTANT: Management endpoints can only be declared by extensions and not from the application code.

== Management Interface Configuration

include::{generated-dir}/config/quarkus-management-management-management-interface-build-time-config.adoc[leveloffset=+1, opts=optional]

include::{generated-dir}/config/quarkus-management-management-management-interface-configuration.adoc[leveloffset=+1, opts=optional]

[[reverse-proxy]]
== Running behind a reverse proxy

> Quarkus could be accessed through proxies that additionally generate headers (e.g. `X-Forwarded-Host`) to keep

Reviewer: ```suggestion
Quarkus can be accessed through proxies that generate headers (e.g. `X-Forwarded-Host`) to preserve
```

> information from the client-facing side of the proxy servers that is altered or lost when they are involved.

Reviewer: ```suggestion
information about the original request.
```

> In those scenarios, Quarkus can be configured to automatically update information like protocol, host, port and URI

Reviewer: ```suggestion
Quarkus can be configured to automatically update information like protocol, host, port and URI to 
```

> reflecting the values in these headers.

Reviewer: ```suggestion
use the values from those headers.
```

> IMPORTANT: Activating this feature leaves the server exposed to several security issues (i.e. information spoofing).

Reviewer: ```suggestion
IMPORTANT: Activating this feature can expose the server to security issues like information spoofing.
```

> Consider activate it only when running behind a reverse proxy.

Reviewer: ```suggestion
Activate it only when running behind a reverse proxy.
```

To set up this feature for the management interface, include the following lines in `src/main/resources/application.properties`:
[source,properties]
----
quarkus.management.proxy.proxy-address-forwarding=true
----

To consider only de-facto standard header (`Forwarded` header), please include the following lines in `src/main/resources/application.properties`:
[source,properties]
----
quarkus.management.proxy.allow-forwarded=true
----

> To consider only non-standard headers, please include the following lines instead in `src/main/resources/application.properties`:

Reviewer: ```suggestion
Alternatively, you can prefer `X-Forwarded-*` headers using the following configuration in `src/main/resources/application.properties` (note `allow-x-forwarded` instead of `allow-forwarded`):
```

[source,properties]
----
quarkus.management.proxy.proxy-address-forwarding=true
quarkus.management.proxy.allow-x-forwarded=true
quarkus.management.proxy.enable-forwarded-host=true
quarkus.management.proxy.enable-forwarded-prefix=true
----

Both configurations related to standard and non-standard headers can be combined, although the standard headers configuration will have precedence. However, combining them has security implications as clients can forge requests with a forwarded header that is not overwritten by the proxy.

> Therefore, proxies should strip unexpected `X-Forwarded` or `X-Forwarded-*` headers from the client.

Reviewer: ```suggestion
Ensure that your proxy is configured to strip unexpected `Forwarded` or `X-Forwarded-*` headers from the client request.
```

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Management interface reference
include::_attributes.adoc[]
:categories: observability
:summary: Management interface configuration
:numbered:
:sectnums:
:sectnumlevels: 4

By default, Quarkus exposes the _management_ endpoints under `/q` on the main HTTP server.
The same HTTP server provides the application endpoints and the management endpoints.

This document presents how you can use a separate HTTP server (bound to a different network interface and port) for the management endpoints.
It avoids exposing these endpoints on the main server and, therefore, prevents undesired accesses.

== Enabling the management interface

To enable the management interface, use the following **build-time** property:

[source, properties]
----

> quarkus.management.enabled=true

Reviewer 2: @cescoffier for clarification — without this option, everything works as it now, and served on `localhost:8080/q` (including, say, arc and Dev UI); and without this option it is served on `${quarkus.management.host}:${q.m.port}${q.m.root-path}`, by default `localhost:9000/q` ? In that case, `quarkus.management.enabled` is probably not the best name, since(from user's perspective), we do not enable something but merely moving it to a separate address.

Author: What property name would you propose?

For the framework it enables the second server.

Reviewer 2: Maybe something like `management.server.enabled`? Or `management.separate-server`?

Author: Yes, we don't expose all /q, just the management endpoints.

----

> Management endpoints will be exposed on: `http://0.0.0.0:9000/q`.

Reviewer 3: ```suggestion
By default, management endpoints will be exposed on: `http://0.0.0.0:9000/q`.
```

For example, `http://0.0.0.0:9000/q/health/ready` for the readiness probe.

SmallRye Health Checks, SmallRye Metrics, and Micrometer Prometheus will be declared as management endpoints when the management interface is enabled.

== Configure the host, port and scheme

By default, the management interface is exposed on the interface: `0.0.0.0` (all interfaces) and on the port `9000` (`9001` in test mode).
It does not use TLS (`https`) by default.

You can configure the host, ports, and TLS certificates using the following properties:

* `quarkus.management.host` - the interface / host
* `quarkus.management.port` - the port
* `quarkus.management.test-port` - the port to use in test mode
* `quarkus.management.ssl` - the TLS configuration, xref:http-reference#ssl[same as for the main HTTP server].

> Here is an example for properties exposing the management interface on _https://localhost:9002_:

Reviewer 3: ```suggestion
Here is a configuration example exposing the management interface on _https://localhost:9002_:
```
